### PR TITLE
Mouse events

### PR DESCRIPTION
**Pull request by Xaeroxe**
Replaces https://github.com/tomaka/winit/pull/192

- Resolves conflicts from that PR
- Adds support for Windows and OSX

### CODE REVIEW

File: src/events.rs [hunk 1]
```diff
@@ -110,10 +113,25 @@ pub enum WindowEvent {
...
+        delta: (f64, f64),
+    },
+
+    /// Physical scroll event
+    MouseWheel {
+        delta: MouseScrollDelta,
+    },
+
+    /// Motion on some analog axis. May report data redundant to other, more specific events.
     Motion { axis: AxisId, value: f64 },
```
Comment (torkleyy) [inline src/events.rs]: Not sure if I understood this. Does this mean that whether `Motion` events are also emitted for the mouse is platform-dependent?

**Comment by torkleyy** [comment]
@Ralith I prefer the names that @Xaeroxe added since `AxisMoved` sounds a bit odd to me since it's not the axis that got moved.

File: src/events.rs (hunk 1 above)
Comment (Xaeroxe) [reply src/events.rs]: Mouse events should be emitted on all major desktop platforms.  Mouse events will also be accompanied by Motion events from the same device with the same deltas.  Not all Motion events will be accompanied by Mouse events though, as some Motion events won't come from the Mouse.

Comment (vberger) [reply src/events.rs]: You could improve this comment to clearly note that `Motion` is a fallback event for any motion along axis that are not already handled by `MouseMoved` or `MouseWheel` (since it seems to be what your implementation does).

**Comment by vberger** [comment]
Looks good to me modulo the doc nitpick. 👍

File: src/events.rs (hunk 1 above)
Comment (Xaeroxe) [reply src/events.rs]: @vberger Oh hm woops, my Windows and OSX implementation of this behaves differently from the X11 implementation provided by @Ralith .  I'm not using the motion events as a fall back, but I'm instead using them to accompany the mouse events.  I'll need to get those behaviors in sync before we merge.

Comment (vberger) [reply src/events.rs]: Ah, I see. So, the actual intended behavior is for the generic `Motion` to repeat `MouseMoved`? Is there a clear reason for this repetition?

Comment (Xaeroxe) [reply src/events.rs]: `Motion` is good for being abstract over axes, so it makes sense to include mouse axes in that abstraction.  `MouseMoved` is for when you don't want to be abstract over axes, and you just want the mouse.

Comment (torkleyy) [reply src/events.rs]: Would be nice to include a detailed description of this in the docs given the confusion in these reviews.

**Comment by Ralith** [comment]
@torkleyy
> I prefer the names that @Xaeroxe added since AxisMoved sounds a bit odd to me because it's not the axis that got moved.

I don't really care what the names are, it'd just be nice if they were consistent.

**Comment by torkleyy** [comment]
@Ralith I'd like them to be consistent, too, is there any other place where `AxisMoved` has been used? Or is it about the `Moved` postfix?

**Comment by Ralith** [comment]
The issue I was discussing was inconsistent use of "Moved" versus "Motion." I don't see a need to have both terms; the introduction of a second one was unintended.

**Comment by torkleyy** [comment]
I see. Let's rename all the variants to `*Motion` then.

**Comment by tomaka** [comment]
If we're breaking the API of `Events`, I think we should group all the changes concerning this enum in the same version in order to reduce friction.

**Comment by vberger** [comment]
looks like it needs a rebase. Likely a consequence of #342 which move the pointer logic for wayland into its own file.

**Comment by Xaeroxe** [comment]
@vberger I've resolved the conflict.

So what's the policy for merging PRs in this repo?

File: src/platform/linux/wayland/event_loop.rs [hunk 2]
```diff
@@ -476,20 +476,136 @@ fn seat_implementation() -> wl_seat::Implementation<SeatIData> {
...
+}
+
+struct PointerIData {
+    sink: Arc<Mutex<EventsLoopSink>>,
+    windows_token: StateToken<WindowStore>,
+    mouse_focus: Option<WindowId>,
+    axis_buffer: Option<(f32, f32)>,
+    axis_discrete_buffer: Option<(i32, i32)>,
+    axis_state: TouchPhase,
+}
```
Comment (vberger) [inline src/platform/linux/wayland/event_loop.rs]: Is it github messing up the displayed diff? It looks like you're undoing part of #342 by re-adding pointer logic in `event_loop.rs` and removing touch support in the seat...

Comment (Xaeroxe) [reply src/platform/linux/wayland/event_loop.rs]: Woops I think I grabbed the wrong pieces while resolving this, thanks and good catch.  Let me try again.

**Comment by Xaeroxe** [comment]
@vberger Ok I think I got it right this time.  git has a really weird idea of how changes impact sections of code sometimes and it confused me because I was expecting git to be smarter than it was.

**Comment by vberger** [comment]
Looks like that's not yet it, the `struct PointerIData` and `fn pointer_implementation()` should not end up in `event_loop.rs`.

I think git is getting confused because you modified some part of this code (while changing the variant names), but #342 has moved this code to an other file (`pointer.rs`)

**Comment by Xaeroxe** [comment]
@vberger Thanks for describing the changes to me, sorry I should have reviewed the PR.  I think this is ready now.

**Comment by vberger** [comment]
Oh, actually one last thing is missing: could you add a line in the changelog describing the breaking changes which were made?

**Comment by Xaeroxe** [comment]
@vberger done.

**Comment by vberger** [comment]
Alright, perfect! 👍
